numpy/core/src/multiarray/item_selection.c

```diff
@@ -787,255 +787,287 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
-    npy_intp astride;
-    PyArray_SortFunc *sort = NULL;
+    npy_intp size;
+
+    int ret = -1;
 
     NPY_BEGIN_THREADS_DEF;
 
+    /* Check if there is any sorting to do */
+    if (N <= 1) {
+        return 0;
+    }
+
     it = (PyArrayIterObject *)PyArray_IterAllButAxis((PyObject *)op, &axis);
-    swap = !PyArray_ISNOTSWAPPED(op);
     if (it == NULL) {
         return -1;
     }
+    size = it->size;
 
     NPY_BEGIN_THREADS_DESCR(PyArray_DESCR(op));
-    if (part == NULL) {
-        sort = PyArray_DESCR(op)->f->sort[swhich];
-    }
-
-    size = it->size;
-    N = PyArray_DIMS(op)[axis];
-    elsize = PyArray_DESCR(op)->elsize;
-    astride = PyArray_STRIDES(op)[axis];
 
-    needcopy = !(PyArray_FLAGS(op) & NPY_ARRAY_ALIGNED) ||
-                (astride != (npy_intp) elsize) || swap;
     if (needcopy) {
-        char *buffer = PyDataMem_NEW(N*elsize);
+        buffer = PyDataMem_NEW(N * elsize);
         if (buffer == NULL) {
             goto fail;
         }
+    }
 
-        while (size--) {
-            _unaligned_strided_byte_copy(buffer, (npy_intp) elsize, it->dataptr,
-                                         astride, N, elsize);
-            if (swap) {
-                _strided_byte_swap(buffer, (npy_intp) elsize, N, elsize);
-            }
-            if (part == NULL) {
-                if (sort(buffer, N, op) < 0) {
-                    PyDataMem_FREE(buffer);
-                    goto fail;
+    while (size--) {
+        char *bufptr = it->dataptr;
+
+        if (needcopy) {
+            if (hasrefs) {
+                /*
+                 * For dtype's with objects, copyswapn Py_XINCREF's src
+                 * and Py_XDECREF's dst. This would crash if called on
+                 * an unitialized buffer, or leak a reference to each
+                 * object if initialized.
+                 *
+                 * So, first do the copy with no refcounting...
+                 */
+                _unaligned_strided_byte_copy(buffer, elsize,
+                                             it->dataptr, astride, N, elsize);
+                /* ...then swap in-place if needed */
+                if (swap) {
+                    copyswapn(buffer, elsize, NULL, 0, N, swap, op);
                 }
             }
             else {
-                npy_intp pivots[NPY_MAX_PIVOT_STACK];
-                npy_intp npiv = 0;
-                npy_intp i;
-                for (i = 0; i < nkth; i++) {
-                    if (part(buffer, N, kth[i], pivots, &npiv, op) < 0) {
-                        PyDataMem_FREE(buffer);
-                        goto fail;
-                    }
-                }
+                copyswapn(buffer, elsize, it->dataptr, astride, N, swap, op);
             }
-            if (swap) {
-                _strided_byte_swap(buffer, (npy_intp) elsize, N, elsize);
+            bufptr = buffer;
+        }
+        else if (swap) {
+            /* If the array is swapped but contiguous, unswap in-place */
```

Comment:
Hmmm... I just realized this optimization could leave a contiguous, swapped array, unwapped if there was an error in the sorting routine, which is not good, will push a new commit to fix it.